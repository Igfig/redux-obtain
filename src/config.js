export let config = {
    requestHeaderSelector: state => ({
        "Content-Type": "application/json",
        Authorization: `Basic ${state.authentication.token}`
    }),
    reduxStoreName: "resources",
    paginationInitialLoadLimit: 500,
    getOrderBys: props =>
        props && props.sortBy && props.sortDirection
            ? props.sortBy.map((column, index) => ({
                  column,
                  direction: props.sortDirection[index]
              }))
            : []
}

export const setupFetcher = options => {
    config = {
        ...config,
        ...options
    }
}
