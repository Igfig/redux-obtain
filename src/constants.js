export const initialState = {}
export const initialResourceState = {
    data: {},
    loading: false,
    error: false
}

const prefix = "@@fetcher"

export const ADD_RESOURCE = `${prefix}/ADD_RESOURCE`
export const MODIFY_RESOURCE = `${prefix}/MODIFY_RESOURCE`
export const REQUEST_RESOURCE = `${prefix}/REQUEST_RESOURCE`
export const FETCH_SUCCESS = `${prefix}/FETCH_SUCCESS`
export const FETCH_ADDITIONAL_SUCCESS = `${prefix}/FETCH_ADDITIONAL_SUCCESS`
export const FETCH_ERROR = `${prefix}/FETCH_ERROR`
export const REMOVE_RESOURCE = `${prefix}/REMOVE_RESOURCE`
