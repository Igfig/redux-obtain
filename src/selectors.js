import { config } from "./config"

export const getResourceData = (name, selector) => state =>
    state[config.reduxStoreName][name]
        ? selector
          ? selector(state[config.reduxStoreName][name].data)
          : state[config.reduxStoreName][name].data
        : undefined

export const getMetaData = name => state => ({
    loading: state[config.reduxStoreName][name].loading,
    error: state[config.reduxStoreName][name].error
})
