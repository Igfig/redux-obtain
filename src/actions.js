import * as C from "./constants"

export const addResource = (name, paginationKey) => ({
    type: C.ADD_RESOURCE,
    name,
    paginationKey
})

export const modifyResource = ({ name, dataTransform }) => ({
    type: C.MODIFY_RESOURCE,
    name,
    dataTransform
})

export const requestResource = name => ({ type: C.REQUEST_RESOURCE, name })

export const fetchSuccess = (name, data, acceptResponse) => ({
    type: C.FETCH_SUCCESS,
    name,
    data,
    acceptResponse
})

export const fetchAdditionalSuccess = (name, data) => ({
    type: C.FETCH_ADDITIONAL_SUCCESS,
    name,
    data
})

export const fetchError = (name, error) => ({ type: C.FETCH_ERROR, name, error })

export const removeResource = name => ({ type: C.REMOVE_RESOURCE, name })
