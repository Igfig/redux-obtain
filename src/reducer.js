import * as C from "./constants"

const createReducer = (initialState, handlers) => (state = initialState, action) =>
    handlers.hasOwnProperty(action.type) ? handlers[action.type](state, action) : state

export const reducer = createReducer(C.initialState, {
    [C.ADD_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: {
            ...C.initialResourceState,
            paginationKey: action.paginationKey
        }
    }),
    [C.MODIFY_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: state[action.name] && {
            ...state[action.name],
            data: action.dataTransform(state[action.name] && state[action.name].data)
        }
    }),
    [C.REQUEST_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            loading: true,
            error: false
        }
    }),
    [C.FETCH_SUCCESS]: (state, action) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: action.acceptResponse ? action.acceptResponse(action.data) : action.data,
            loading: false,
            error: false
        }
    }),
    [C.FETCH_ADDITIONAL_SUCCESS]: (state, action) => {
        const paginationKey = state[action.name].paginationKey
        const newData = action.acceptResponse ? action.acceptResponse(action.data) : action.data
        return {
            ...state,
            [action.name]: {
                ...state[action.name],
                data: {
                    ...newData,
                    [paginationKey]: paginationKey && [
                        ...state[action.name].data[paginationKey],
                        ...action.data[paginationKey]
                    ]
                },
                loading: false,
                error: false
            }
        }
    },
    [C.FETCH_ERROR]: (state, action) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: {},
            loading: false,
            error: true
        }
    }),
    [C.REMOVE_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: undefined
    })
})
