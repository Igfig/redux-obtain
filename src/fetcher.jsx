import React, { Component } from "react"
import { connect } from "react-redux"
import * as actions from "./actions"
import * as C from "./constants"
import { config } from "./config"
import Promise from "bluebird"
import _ from "lodash"
import getDisplayName from "react-display-name"

Promise.config({ cancellation: true })

export const fetcher = (
    {
        name,
        endpoint,
        method,
        acceptResponse,
        persistResource,
        requestBodySelector,
        defaultOrderBys,
        paginationKey
    },
    extraActions
) => WrappedComponent =>
    connect(
        (state, ownProps) => ({
            endpoint: typeof endpoint === "function" ? endpoint(state, ownProps) : endpoint,
            requestHeader: config.requestHeaderSelector(state),
            resource: state[config.reduxStoreName][name],
            requestBody: requestBodySelector && requestBodySelector(state, ownProps)
        }),
        {
            ...extraActions,
            ...actions
        }
    )(
        class Fetcher extends Component {
            static displayName = `Fetcher(${getDisplayName(WrappedComponent)})`
            componentWillMount() {
                this.props.addResource(name, paginationKey)
            }
            componentDidMount() {
                this.sendNetworkRequest({
                    limit: config.paginationInitialLoadLimit,
                    offset: 0,
                    orderBys: config.getOrderBys(defaultOrderBys),
                    firstLoad: true,
                    requestBody: this.props.requestBody
                })
            }
            componentWillUnmount() {
                this.networkRequest && this.networkRequest.cancel()
                !persistResource && this.props.removeResource(name)
            }
            componentWillReceiveProps(nextProps) {
                if (!_.isEqual(this.props.requestBody, nextProps.requestBody)) {
                    this.sendNetworkRequest({
                        limit: config.paginationInitialLoadLimit,
                        offset: 0,
                        orderBys: config.getOrderBys(defaultOrderBys),
                        firstLoad: true,
                        requestBody: nextProps.requestBody
                    })
                }
            }
            sendNetworkRequest = ({ limit, offset, orderBys, firstLoad, requestBody }) => {
                firstLoad && this.props.requestResource(name)
                const paginationBody = paginationKey ? { limit, offset, orderBys } : {}
                const requestMethod = paginationKey ? "POST" : method
                this.networkRequest = new Promise((res, rej) =>
                    fetch(this.props.endpoint, {
                        method: requestMethod,
                        headers: this.props.requestHeader,
                        body:
                            (method !== "GET" || !!paginationKey) &&
                            JSON.stringify({
                                ...requestBody,
                                ...paginationBody
                            })
                    })
                        .then(x => res(x))
                        .catch(e => rej(e))
                )
                    .then(res => {
                        if (res.status === 200) {
                            return res
                                .json()
                                .then(
                                    data =>
                                        firstLoad
                                            ? this.props.fetchSuccess(name, data, acceptResponse)
                                            : this.props.fetchAdditionalSuccess(
                                                  name,
                                                  data,
                                                  acceptResponse
                                              )
                                )
                        } else {
                            return res.text().then(error => this.props.fetchError(name, error))
                        }
                    })
                    .catch(e => console.error(e))
                return this.networkRequest
            }
            loadMoreRows = (indices, ui) => {
                const { startIndex, stopIndex } = indices
                return this.sendNetworkRequest({
                    limit: stopIndex - startIndex,
                    offset: startIndex,
                    orderBys: config.getOrderBys(ui),
                    requestBody: this.props.requestBody,
                    firstLoad: false
                })
            }
            loadInitialRows = props => {
                return this.sendNetworkRequest({
                    limit: config.paginationInitialLoadLimit,
                    offset: 0,
                    orderBys: config.getOrderBys(props),
                    requestBody: this.props.requestBody,
                    firstLoad: true
                })
            }
            render() {
                let pt = { ...this.props }
                delete pt.endpoint
                delete pt.addResource
                delete pt.modifyResource
                delete pt.requestResource
                delete pt.fetchSuccess
                delete pt.fetchAdditionalSuccess
                delete pt.fetchError
                delete pt.removeResource
                delete pt.requestBody
                delete pt.requestHeader
                delete pt.resource
                const paginationFunctions = paginationKey
                    ? {
                          _loadMoreRows: this.loadMoreRows,
                          _loadInitialRows: this.loadInitialRows,
                          totalSize: this.props.resource && this.props.resource.data.totalCount
                      }
                    : undefined
                return this.props.resource ? (
                    <WrappedComponent
                        {...pt}
                        {...this.props.resource}
                        paginationFunctions={paginationFunctions}
                    />
                ) : (
                    <div />
                )
            }
        }
    )
