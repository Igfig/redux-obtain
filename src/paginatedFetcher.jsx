import React, { Component } from "react"
import { connect } from "react-redux"
import * as actions from "./actions"
import * as C from "./constants"
import { config } from "./config"
import Promise from "bluebird"
import _ from "lodash"
import getDisplayName from "react-display-name"

Promise.config({ cancellation: true })

const getOrderBys = ({ sortBy, sortDirection }) =>
    sortBy &&
    sortDirection &&
    sortBy.map((column, index) => ({ column, direction: sortDirection[index] }))

export const paginatedFetcher = (
    { name, endpoint, defaultOrderBys, persistResource, requestBodySelector, paginationKey },
    extraActions
) => WrappedComponent =>
    connect(
        (state, ownProps) => ({
            endpoint: typeof endpoint === "function" ? endpoint(state, ownProps) : endpoint,
            requestHeader: config.requestHeaderSelector(state),
            resource: state[config.reduxStoreName][name],
            requestBody: requestBodySelector && requestBodySelector(state, ownProps)
        }),
        {
            ...extraActions,
            ...actions
        }
    )(
        class PaginatedFetcher extends Component {
            static displayName = `PaginatedFetcher(${getDisplayName(WrappedComponent)})`
            componentWillMount() {
                this.props.addResource(name, paginationKey)
            }
            componentDidMount() {
                this.sendNetworkRequest({
                    limit: config.paginationInitialLoadLimit,
                    offset: 0,
                    orderBys: getOrderBys(defaultOrderBys),
                    firstLoad: true,
                    requestBody: this.props.requestBody
                })
            }
            componentWillUnmount() {
                this.networkRequest && this.networkRequest.cancel()
                !persistResource && this.props.removeResource(name)
            }
            componentWillReceiveProps(nextProps) {
                if (!_.isEqual(this.props.requestBody, nextProps.requestBody)) {
                    this.sendNetworkRequest({
                        limit: config.paginationInitialLoadLimit,
                        offset: 0,
                        orderBys: getOrderBys(defaultOrderBys),
                        firstLoad: true,
                        requestBody: nextProps.requestBody
                    })
                }
            }
            sendNetworkRequest = ({ limit, offset, orderBys, firstLoad, requestBody }) => {
                firstLoad && this.props.requestResource(name)
                this.networkRequest = new Promise((res, rej) =>
                    fetch(this.props.endpoint, {
                        method: "POST",
                        headers: this.props.requestHeader,
                        body: JSON.stringify({
                            limit,
                            offset,
                            orderBys,
                            ...requestBody
                        })
                    })
                        .then(x => res(x))
                        .catch(e => rej(e))
                )
                    .then(res => {
                        if (res.status === 200) {
                            return res.json()
                        } else {
                            res.text()
                            throw new Error(`Web Service responded with ${res.status}`)
                        }
                    })
                    .then(
                        data =>
                            firstLoad
                                ? this.props.fetchSuccess(name, data)
                                : this.props.fetchAdditionalSuccess(name, data)
                    )
                    .catch(error => {
                        console.log(error)
                        return this.props.fetchError(name, error)
                    })
                return this.networkRequest
            }
            loadMoreRows = (indices, ui) => {
                const { startIndex, stopIndex } = indices
                return this.sendNetworkRequest({
                    limit: stopIndex - startIndex,
                    offset: startIndex,
                    orderBys: getOrderBys(ui),
                    requestBody: this.props.requestBody,
                    firstLoad: false
                })
            }
            loadInitialRows = props => {
                return this.sendNetworkRequest({
                    limit: config.paginationInitialLoadLimit,
                    offset: 0,
                    orderBys: getOrderBys(props),
                    requestBody: this.props.requestBody,
                    firstLoad: true
                })
            }
            render() {
                let pt = { ...this.props }
                delete pt.addResource
                delete pt.modifyResource
                delete pt.requestResource
                delete pt.fetchSuccess
                delete pt.fetchAdditionalSuccess
                delete pt.fetchError
                delete pt.removeResource
                delete pt.requestBody
                delete pt.requestHeader
                delete pt.resource
                return this.props.resource ? (
                    <WrappedComponent
                        {...pt}
                        {...this.props.resource}
                        paginationFunctions={{
                            _loadMoreRows: this.loadMoreRows,
                            _loadInitialRows: this.loadInitialRows,
                            totalSize: this.props.resource.data.totalCount
                        }}
                    />
                ) : (
                    <div />
                )
            }
        }
    )
