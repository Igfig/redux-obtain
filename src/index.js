import { fetcher } from "./fetcher"
import { modifyResource } from "./actions"
import { getResourceData, getMetaData } from "./selectors"
import { reducer } from "./reducer"
import { setupFetcher } from "./config"

export { fetcher, modifyResource, getResourceData, getMetaData, reducer, setupFetcher }
