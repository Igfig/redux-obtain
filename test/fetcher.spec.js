import React, { Component } from "react"
import { Provider } from "react-redux"
import { createStore, combineReducers } from "redux"
import { shallow, mount } from "enzyme"
import toJson from "enzyme-to-json"
import { fetcher, reducer } from "../src"
import configureStore from "redux-mock-store"
const mockStore = configureStore()
Date.now = jest.fn(() => 42)
global.fetch = require("jest-fetch-mock")

const initialStore = {
    authentication: { token: "dummy_token" },
    resources: {}
}

describe("Fetcher", () => {
    it("should mount, call fetch and dispatch actions", () => {
        const Container = fetcher({ name: "NAME", endpoint: "/api/endpoint", method: "GET" })(
            () => <div />
        )
        const store = mockStore(initialStore)
        const wrapper = mount(
            <Provider store={store}>
                <Container />
            </Provider>
        )
        expect(toJson(wrapper)).toMatchSnapshot()
        expect(store.getActions()).toMatchSnapshot()
    })
    it("should dispatch unmount action on unmount", () => {
        const Container = fetcher({ name: "NAME", endpoint: "/api/endpoint", method: "GET" })(
            () => <div />
        )
        const store = mockStore(initialStore)
        const wrapper = mount(
            <Provider store={store}>
                <Container />
            </Provider>
        )
        wrapper.unmount()
        expect(store.getActions()).toMatchSnapshot()
    })
})
