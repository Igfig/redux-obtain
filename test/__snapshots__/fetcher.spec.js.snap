// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetcher should dispatch unmount action on unmount 1`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REMOVE_RESOURCE",
  },
]
`;

exports[`Fetcher should mount, call fetch and dispatch actions 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Fetcher(Unknown))>
    <Fetcher(Unknown)
      addResource={[Function]}
      endpoint="/api/endpoint"
      fetchAdditionalSuccess={[Function]}
      fetchError={[Function]}
      fetchSuccess={[Function]}
      modifyResource={[Function]}
      removeResource={[Function]}
      requestHeader={
        Object {
          "Authorization": "Basic dummy_token",
          "Content-Type": "application/json",
        }
      }
      requestResource={[Function]}
    >
      <div />
    </Fetcher(Unknown)>
  </Connect(Fetcher(Unknown))>
</Provider>
`;

exports[`Fetcher should mount, call fetch and dispatch actions 2`] = `
Array [
  Object {
    "name": "NAME",
    "paginationKey": undefined,
    "type": "@@fetcher/ADD_RESOURCE",
  },
  Object {
    "name": "NAME",
    "type": "@@fetcher/REQUEST_RESOURCE",
  },
]
`;
